cmd/kubeadm/app/util/staticpod/utils.go

```diff
@@ -140,3 +145,65 @@ func WriteStaticPodToDisk(componentName, manifestDir string, pod v1.Pod) error {
 
 	return nil
 }
+
+// GetProbeAddress returns an IP address or 127.0.0.1 to use for liveness probes
+// in static pod manifests. For etcd, GetProbeAddress will use the first
+// listen-client-url if multiple urls are provided. If the listen-client-url
+// contains a hostname instead of an IP address, GetProbeAddress will use the
+// local resolver for name resolution or 127.0.0.1 if the name can not be resolved.
+func GetProbeAddress(cfg *kubeadmapi.MasterConfiguration, componentName string) string {
+	switch {
+	case componentName == kubeadmconstants.KubeAPIServer:
+		if cfg.API.AdvertiseAddress != "" {
+			return cfg.API.AdvertiseAddress
+		}
+	case componentName == kubeadmconstants.KubeScheduler:
+		if cfg.SchedulerExtraArgs != nil {
+			addr, exists := cfg.SchedulerExtraArgs["address"]
+			if exists {
+				return addr
+			}
+		}
+	case componentName == kubeadmconstants.KubeControllerManager:
+		if cfg.ControllerManagerExtraArgs != nil {
+			addr, exists := cfg.ControllerManagerExtraArgs["address"]
+			if exists {
+				return addr
+			}
+		}
+	case componentName == kubeadmconstants.Etcd:
+		if cfg.Etcd.ExtraArgs != nil {
+			arg, exists := cfg.Etcd.ExtraArgs["listen-client-urls"]
+			if exists {
+				if strings.ContainsAny(arg, ",") {
+					arg = strings.Split(arg, ",")[0]
+				}
+				parsedURL, err := url.Parse(arg)
+				if err != nil || parsedURL.Hostname() == "" {
+					break
+				}
+				if ip := net.ParseIP(parsedURL.Hostname()); ip != nil {
+					return ip.String()
+				}
+				addrs, err := net.LookupIP(parsedURL.Hostname())
+				if err != nil {
+					break
+				}
+				var ip net.IP
+				for _, addr := range addrs {
+					if addr.To4() != nil {
+						ip = addr
+						break
+					}
+					if addr.To16() != nil && ip == nil {
+						ip = addr
```

Comment:
seems like you could do `addr.To16().String()` here to make the code cleaner?

Reply:
ping @danehans

Reply:
@luxas I updated the PR to do `addr.To16().String()` and the same for To4, but I don't think that is exactly the approach we want to take. We want to iterate over the slice of IP's and prefer a v4 address over a v6 address. In the above, we want to assign the v6 addr to ip to save the value and loop back though the slice. This is b/c the next ip may be a v4 ip. I am going to push a new patch set. Take a look and let me know if it makes sense to you.